packages/orderbook/src/order_provider/mesh_order_provider.ts

```diff
@@ -0,0 +1,207 @@
+import { APIOrder } from '@0x/connect';
+import {
+    AcceptedOrderInfo,
+    OrderEvent,
+    OrderEventKind,
+    OrderInfo,
+    RejectedOrderInfo,
+    WSClient,
+} from '@0x/mesh-rpc-client';
+import { Asset, AssetPairsItem, SignedOrder } from '@0x/types';
+import { BigNumber } from '@0x/utils';
+
+import { OrderStore } from '../order_store';
+import { AcceptedRejectedOrders, AddedRemovedOrders, MeshOrderProviderOpts } from '../types';
+import { utils } from '../utils';
+
+import { BaseOrderProvider } from './base_order_provider';
+
+export class MeshOrderProvider extends BaseOrderProvider {
+    private readonly _wsClient: WSClient;
+    private _wsSubscriptionId?: string;
+
+    /**
+     * Converts the OrderEvent or OrderInfo from Mesh  into an APIOrder.
+     * If the OrderInfo is a RejectedOrderInfo the remainingFillableTakerAssetAmount is
+     * assumed to be 0.
+     * @param orderEvent The `OrderEvent` from a Mesh subscription update
+     */
+    private static _orderInfoToAPIOrder(
+        orderEvent: OrderEvent | AcceptedOrderInfo | RejectedOrderInfo | OrderInfo,
+    ): APIOrder {
+        const remainingFillableTakerAssetAmount = (orderEvent as OrderEvent).fillableTakerAssetAmount
+            ? (orderEvent as OrderEvent).fillableTakerAssetAmount
+            : new BigNumber(0);
+        return {
+            order: orderEvent.signedOrder,
+            metaData: {
+                orderHash: orderEvent.orderHash,
+                remainingFillableTakerAssetAmount,
+            },
+        };
+    }
+
+    /**
+     * Instantiates a [Mesh](https://github.com/0xProject/0x-mesh) Order Provider. This provider writes
+     * all orders stored in Mesh to the OrderStore and subscribes all Mesh updates.
+     * @param opts `MeshOrderProviderOpts` containing the websocketEndpoint and additional Mesh options
+     * @param orderStore The `OrderStore` where orders are added and removed from
+     */
+    constructor(opts: MeshOrderProviderOpts, orderStore: OrderStore) {
+        super(orderStore);
+        this._wsClient = new WSClient(opts.websocketEndpoint, opts.wsOpts);
+    }
+
+    /**
+     * Returns the available asset pairs. If no subscription to Mesh exists (and therefor no orders) it is
+     * created and awaited on. Once the connection has been initialized the orders in the store are returned
+     * as asset pairs.
+     */
+    public async getAvailableAssetDatasAsync(): Promise<AssetPairsItem[]> {
+        await this._initializeIfRequiredAsync();
+        const assetPairsItems: AssetPairsItem[] = [];
+        const minAmount = new BigNumber(0);
+        const maxAmount = new BigNumber(2).pow(256).minus(1);
+        const precision = 18;
```

Comment:
Hard-coding `18` precision here will wreck havoc no?

Reply:
Precision isn't decimals, it's what the Relayer allows on their orderbook, i.e 5 precision on 18 decimals. Since we're reusing the interface we need a value and I don't believe Mesh has any limits on precision?